include: devicetree: Generate defines for map property

soburi opened the pull request:
This change introduces generating definitions corresponding to
`*-map` property, which was currently discarded.

The `*-map` data is like a two-dimensional array, so it is difficult to
handle with the existing APIs, so we will also provide new APIs.

---

Common utilities have been separated into a separate PR.
Please take a look at this first. https://github.com/zephyrproject-rtos/zephyr/pull/84190


---

I noticed this problem while developing https://github.com/zephyrproject-rtos/gsoc-2022-arduino-core.

Since map is used to associate with the physical layout, there is a need to know the "physical pin names".

Since the mapping process is performed during the generation of devicetree-generated.h, I don't think there are many cases where information about the mapping source is needed. However, there are such cases, however rare, and it is not desirable for the information written in the devicetree to be discarded, so I would like to introduce this.

soburi committed:
sys: util: Add `GET_ARGS_FIRST_N` to get the first N of the args

Gets the specified number of elements from the beginning of the list.
This is the symmetircal operation as `GET_ARGS_LESS_N`.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
tests: unit: util: Add test for `GET_ARGS_FIRST_N`

Add `GET_ARGS_FIRST_N` tests.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
scripts: dts: Add handling for `*-map` property

This change introduces generating definitions corresponding to
`*-map` property, which was currently discarded.

For `*-map` properties are made able to be treated as a variation
of phandle-array, assign sequential cell names for each group of
specifiers (child_specifier_0, child_specifier_1, ...,
parent_specifier_0, ...).

The `*-map` data is like a two-dimensional array, so it is difficult to
handle with the existing APIs, so we will also provide new APIs.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
dts: bindings: Change `*-map-mask` and `*-map-pass-thru` as array type.

They must be `array` type to match the existing `interrupt-map-mask`
and `interrupt-map-pass-thru`.

And also, updating descriptions.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
dts: bindings: Add interrupt-nexus and io-channel-nexus base bindings

Add a base binding for interrupt and io-channel nexus nodes.
This makes the implicit definitions explicit.
Replace existing individual map definitions with this.

This file does not define a specific `compatible` on its own,
So you should include this file before using it practically.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

mbolivar commented on dts/bindings/adc/io-channel-nexus.yaml:
```diff
@@ -0,0 +1,20 @@
+  "#io-channel-cells":
+    type: int
+    required: true
+    description: Number of items to expect in a ADC specifier
```
I think, strictly speaking, that these are derived from the Linux iio bindings. The iio bindings are useful for ADCs but are not limited to ADCs. So perhaps this should be moved to dts/bindings/iio, and this "ADC specifier" portion should be changed to "channel specifier, e.g. ADC channel" or something similar?

soburi replied:
That's correct. We've taken care of that.

mbolivar reviewed (commented):
Thanks for the PR, @soburi . The first 5/8 patches look good to me, with one minor comment.

I have some concerns about the last 3/8 patches. I think I understand the goal and exposing the map properties as first-class citizens makes sense to me. However, the approach using `to_compound()` feels a little bit error-prone and I want to discuss the implementation.

Devicetree Specification v0.4 section 2.5, "Nexus Nodes and Specifier Mapping" is where the semantics are defined. There, it says:

> A nexus node shall have a `#<specifier>-cells` property, where `<specifier>` is some specifier space such as ‘gpio’, ‘clock’, ‘reset’, etc

The description of the `<specifier>-map` property then says:

> child specifier
>
> The specifier of the child node being mapped. The number of 32-bit cells required to specify this component is described by the `#<specifier>-cells` property of this node—the nexus node containing the `<specifier>-map` property

It also says:

> parent specifier
>
> The specifier in the parent domain. The number of 32-bit cells required to specify this component is described by the `#<specifier>-cells` property of the specifier parent node.

Combining these, I think that we should be able to decode each `<specifier>-map` property without needing `to_compound()`. We can make a loop like this:

1. Read `#<specifier>-cells * 4` bytes from the current position in the `<specifier>-map` property. This is the child specifier. We can convert it to `<u32>` cells using existing dtlib features.
2. Read the next 4 bytes. This is the "specifier parent" phandle from section 2.5 of the specification. We can convert it to a "specifier parent node" using phandle2node.
3. Read the `#<specifier>-cells` property from the specifier parent node. This is the number of `<u32>` cells after the phandle which makes the "parent specifier"
4. Read the parent specifier bytes and convert them to <u32> cells.

What I mean is, by using the `#<specifier>-cells` properties in the parent and child nodes, we can iterate through the `<specifier>-map` property and convert it into a lookup table without `to_compound()`.

I am concerned that `to_compound()` is:

- not checking the proper `#<specifier>-cells` values we expect to find, and instead just uses the source encoding to guess where the table entry elements are in `<specifier>-map`
- adding a feature for strings that does not seem to be needed to handle `<speciifer>-map` properties -- this feels like it might be trying to do more than it needs to, and might have to be revisited over time as we add more heuristics

I'd be more comfortable with an approach that relied on the strict definitions in section 2.5 of the specification to decode the values of `<specifier>-map` properties. We can be sure that this will not change and  the code will stand the test of time.

soburi commented:
@mbolivar 

> Thanks for the PR, @soburi . The first 5/8 patches look good to me, with one minor comment.
> 
> I have some concerns about the last 3/8 patches. I think I understand the goal and exposing the map properties as first-class citizens makes sense to me. However, the approach using `to_compound()` feels a little bit error-prone and I want to discuss the implementation.
> 
> Devicetree Specification v0.4 section 2.5, "Nexus Nodes and Specifier Mapping" is where the semantics are defined. There, it says:
> 

Thank you for your review.

I was also thinking about the implementation policy here, so
I would be happy to discuss it.

Before going into the implementation details, let me discuss a design-level topic.

I understand that the current implementation is designed as follows:

- dtlib.py: Responsible for grammar processing
- edtlib.py: Responsible for semantic interpretation

IMO, In this division of roles, the number of parameters specified in #gpio-cells is an issue in the "semantic" domain, so I think it would be better to implement the check in edtlib.py.

In that case, in the domain of dtlib.py, this is 'compound' as a syntax element, and even if string handling is left as 'not implemented', I think there should be a 'to_compound' method here.

I believe that once this division of roles is made clear, the implementation will become clear naturally, so I would like to get this issue resolved first.

mbolivar commented:
@soburi thanks for your update. I agree we should discuss this design-level issue first.

The "compound" concept appears in both edtlib and dtlib:

- in edtlib, we have the "compound" type in the bindings syntax, which I believe historically was added mainly to document and capture complex cases that we couldn't express in our existing syntax
- in dtlib, the "compound" type in the DTS syntax parser exists as a catch-all for properties that we don't have special-case support for. It's important to remember that from the perspective of the DT specification, all properties are just byte arrays, and the source encoding was ignored by the consumer of the DTB as defined in chapter 5 of the specification. Zephyr is unique among operating systems in that it reads the DTS directly, of course. The dtlib parser has several special case "types" encoded in the dtlib.Type enum. For properties that don't fit in any special case, we can still access the byte array value, and the property is exposed as type dtlib.Type.COMPOUND.

I will try to be clear about which case I am talking about when we think about nexus nodes:

- in edtlib, our existing bindings syntax doesn't have a concept of a nexus node and so the only way we can name such properties in the bindings is with type 'compound' in the bindings syntax
- in dtlib, we are specifically concerned with the properties defined in chapter 2.5 of the specification, whose format and syntax fits into these dtlib types:

  - `<specifier>-map`: the specification describes properties of dtlib.Type.PHANDLES_AND_NUMS, not dtlib.Type.COMPOUND
  - `<specifier>-map-mask`: dtlib.Type.NUMS
  - `<specifier>-map-pass-thru`: dtlib.Type.NUMS
  - `#<specifier>-cells`: dtlib.Type.NUM

I think that since our dtlib parser already seems to have the right dtlib.Type special cases to handle nexus nodes, we probably don't need any dtlib changes to handle these. However, since the edtlib library doesn't have support for this pattern in its bindings syntax, I think we have two choices:

1. extend the bindings syntax to add new special case types for `<specifier>-map`
2. add special-case handling for this chapter of the devicetree specification, similarly to how we handled GPIO hogs

I think choice 2 is best, because this is a special case that is defined directly in the specification. We already allow edtlib to handle these types of special cases for things like the interrupt tree (DT specification section 2.4), major properties like `reg` and `compatible` (specification section 2.3), etc.

I outlined a way that I think we can implement option 2 that checks that the user's devicetree matches the requirements of specification section 2.5, Nexus Nodes and Specifier Mapping, in my previous review comment.

Does it make sense? What are your thoughts?

soburi commented:
@mbolivar 

Thank you for getting back to me.

Indeed, if I do it like this, I can use dtlib.py as it is.

```
diff --git a/dts/bindings/gpio/gpio-nexus.yaml b/dts/bindings/gpio/gpio-nexus.yaml
index de8153b0f9a..311e4cfb1cd 100644
--- a/dts/bindings/gpio/gpio-nexus.yaml
+++ b/dts/bindings/gpio/gpio-nexus.yaml
@@ -5,7 +5,7 @@

properties:
gpio-map:
- type: compound
+ type: phandle-array
required: true

gpio-map-mask:
```

It seems like it would be good to follow your suggestion in general,
but I will reconsider it based on this. Thanks.

soburi committed:
tests: lib: devicetree: api: Add map property tests

Add tests for map property related macros.

Add `native_sim/native/64` to allowed platforms.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented:
@mbolivar 

I finally solved the problem by adding a maps @property to edtlib.
Could you review again, please?

soburi commented:
@mbolivar 
ping

rruuaanng commented on include/zephyr/devicetree/map.h:
```diff
@@ -0,0 +1,263 @@
+ * @param idx The mapping entry index.
+ * @return The mapping entry as a list of comma-separated values.
+ */
+#define DT_MAP_IDX(node_id, prop, idx) DT_CAT5(node_id, _P_, prop, _MAP_IDX_, idx)
```
```suggestion
#define DT_MAP_BY_IDX(node_id, prop, idx) DT_CAT5(node_id, _P_, prop, _MAP_IDX_, idx)
```

Should we continue to stick with the current style? At present, most APIs that take an `idx` parameter use the `_BY_IDX` suffix and also provide a simpler version of the API where `idx` defaults to 0.

soburi replied:
I have revised the whole code to comply with these conventions.

soburi commented:
@rruuaanng @mbolivar 

I have updated it, so please check it out again if you have time.

soburi commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1321,6 +1321,97 @@ def gpio_hogs(self) -> list[ControllerAndData]:
+                raw = raw[4:]
+
+                controller_node = prop.node.dt.phandle2node.get(phandle)
+                if not controller_node:
```
Sorry for my late response.
I fixed it.

soburi commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1321,6 +1321,97 @@ def gpio_hogs(self) -> list[ControllerAndData]:
+                    _err(f"controller node cannot be found from phandle:{phandle}")
+
+                controller: Node = self.edt._node2enode[controller_node]
+                if not controller:
```
Fixed it.

soburi commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1321,6 +1321,97 @@ def gpio_hogs(self) -> list[ControllerAndData]:
+                    # Not enough room for phandle
+                    _err("bad value for " + repr(prop))
+
+                def count_cells_num(node: dtlib_Node, specifier: str) -> int:
```
Fixed it as your suggestion.

soburi commented:
@rruuaanng 

Is it possible to proceed with this PR? If there are any issues, please let me know and I will address them.

rruuaanng reviewed (commented):
Looks good, let's hear the suggestions from the other developers.

soburi committed:
scripts: dts: devicetree: tests: Add a map property test

Add a map property test for `test_edtlib.py`.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
Thank you for your comment. I’ve added the test you suggested.
This PR has been open for quite some time, so there has been enough time for feedback.
If there are no outstanding issues, could you please approve it? If anyone has objections, I’m sure they’ll raise them.

mbolivar reviewed (changes requested):
Quick first pass review -- I very much like the idea. Can you please update the documentation as well so we can track these new macros? I also want to see the macros.bnf update so I can check that the implementation matches it.

soburi commented on tests/unit/util/main.c:
```diff
@@ -527,6 +527,28 @@ ZTEST(util, test_GET_ARGS_LESS_N) {
+	uint8_t a[] = {GET_ARGS_FIRST_N(0, 1, 2, 3)};
+	uint8_t b[] = {GET_ARGS_FIRST_N(1, 1, 2, 3)};
+	uint8_t c[] = {GET_ARGS_FIRST_N(2, 1, 2, 3)};
+	uint8_t d[] = {GET_ARGS_FIRST_N(3, 1, 2, 3)};
```
This will result in a compilation error.
Specifically, the N portion is a non-variable-length argument, so it will be detected as an argument mismatch. This allows for early error detection. (delightful!)

```
/home/crs/zephyrproject/zephyr/tests/unit/util/main.c:568:8: error: macro "Z_GET_ARGS_FIRST_4" requires 5 arguments, but only 3 given
  568 |         uint8_t e[] = {GET_ARGS_FIRST_N(4, 1, 2, 3)};
```

mbolivar commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1321,6 +1321,97 @@ def gpio_hogs(self) -> list[ControllerAndData]:
+            add up the values of the address cells.
+            """
+
+            if node is None:
+                _err("node is None.")
```
This should be impossible if the type checking is working properly. The only way `node` could be `None` is if its type annotation was `dtlib_Node | None`. The fact that the type is `dtlib_Node` already means it cannot be`None`. If that is not true, something is wrong with the type checking or the type annotation.

soburi replied:
Admittedly, this is redundant if there is type checking.
Removed.

mbolivar commented on scripts/dts/python-devicetree/src/devicetree/edtlib.py:
```diff
@@ -1321,6 +1321,97 @@ def gpio_hogs(self) -> list[ControllerAndData]:
         return res
 
+    @property
+    def maps(self) -> list[ControllerAndData]:
```
This is adding a new user-facing API to edtlib without documenting it. Please compare this with the way other properties (like `gpio_hogs`) are written and documented and add the missing documentation.

It should look like this:

```py
    @property
    def maps(self) -> list[ControllerAndData]:
        "See the class docstring"
        ...
```

And then the class docstring for `Node` should be extended to document the `maps` property appropriately.

It's hard for me to check if the function is working correctly if its behavior is not documented. It's also necessary to document all the public APIs so users of this python package know how they are supposed to work.

soburi replied:
Added description. Please check it.

mbolivar commented on doc/build/dts/macros.bnf:
```diff
@@ -55,6 +55,14 @@ node-macro =/ %s"DT_N" path-id %s"_RANGES_IDX_" DIGIT
 node-macro =/ %s"DT_N" path-id %s"_RANGES_IDX_" DIGIT
               %s"_VAL_CHILD_BUS_FLAGS_EXISTS"
 node-macro =/ %s"DT_N" path-id %s"_FOREACH_RANGE"
+; The map specific properties.
```
This list appears to be incomplete. For example, look at this section from `write_maps()`:

```
+    # DT_N_<node-id>_P_<prop-id>_FOREACH_PROP_ELEM
+    macro2val[f"{macro}_FOREACH_PROP_ELEM(fn)"] = ' \\\n\t'.join(
+        f'fn(DT_{node.z_path_id}, {prop_id}, {i})' for i in range(plen)
+    )
+
+    # DT_N_<node-id>_P_<prop-id>_FOREACH_PROP_ELEM_SEP
+    macro2val[f"{macro}_FOREACH_PROP_ELEM_SEP(fn, sep)"] = ' DT_DEBRACKET_INTERNAL sep \\\n\t'.join(
+        f'fn(DT_{node.z_path_id}, {prop_id}, {i})' for i in range(plen)
+    )
+
+    # DT_N_<node-id>_P_<prop-id>_FOREACH_PROP_ELEM_VARGS
+    macro2val[f"{macro}_FOREACH_PROP_ELEM_VARGS(fn, ...)"] = ' \\\n\t'.join(
+        f'fn(DT_{node.z_path_id}, {prop_id}, {i}, __VA_ARGS__)' for i in range(plen)
+    )
+
+    # DT_N_<node-id>_P_<prop-id>_FOREACH_PROP_ELEM_SEP_VARGS
+    macro2val[f"{macro}_FOREACH_PROP_ELEM_SEP_VARGS(fn, sep, ...)"] = (
+        ' DT_DEBRACKET_INTERNAL sep \\\n\t'.join(
+            f'fn(DT_{node.z_path_id}, {prop_id}, {i}, __VA_ARGS__)' for i in range(plen)
+        )
+    )
```

mbolivar replied:
Ah, I see. Clever. Thanks for clarifying.

mbolivar reviewed (changes requested):
@soburi thank you for your update.

The first 4 commits look good to me.

I ran into some issues while reviewing commit number 5, `scripts: dts: Add handling for *-map property`. Please check these comments. Anything marked "nonblocking" is just an idea that won't block the merge from my side.

soburi commented on dts/bindings/iio/io-channel-nexus.yaml:
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2025 TOKITA Hiroshi
+# SPDX-License-Identifier: Apache-2.0
+
+# Common fields for io-channel nexus
```
Applied your suggestion.
Also gpio-nexus.yaml and pwm-nexus.yaml

mbolivar reviewed (commented):
@soburi thank you for your update. I think there is enough in here for me to give this a good review, which I will do as soon as I can. (I am traveling this week on a work trip so my availability will be a bit less than it was last week, but I want to prioritize your PR among other DT review requests I have.)

As an overall design question, I'm not sure that I think that ControllerAndData is an appropriate class to be using to represent the maps.

The zephyr phandle-array type is used when we have a list of (phandle, specifier) pairs. But in the case of generic maps, we don't have pairs, we have triples: (child_specifier, specifier_parent_phandle, parent_specifier). And in the case of interrupt maps, we have 5-tuples: (child_unit_address, child_interrupt_specifier, interrupt_parent, parent_unit_address, parent_interrupt_specifier).

I think that neither of these types of information is really a ControllerAndData due to the additional "child -> parent" information.

I am currently thinking it would make more sense to introduce two new types:

1. `NexusMapEntry`: for the table entries contained in a `<specifier>-map` as documented in DTSpec section 2.5.1. This would contain attributes named `child_specifier`, `parent`, `parent_specifier`.
2. `InterruptMapEntry`: similar, except for the `interrupt-map` property table entries in DTSpec 2.4.3.

This would match the same pattern we have done with the `Range` type used to represent elements of the standard `ranges` property. Experience has shown from that case that picking type names and designing APIs to match the specification leads to future-proof work.

What do you think about this idea?

mbolivar commented:
> There may be some room for debate as to whether to use `NexusMapEntry`/`InterruptMapEntry` or use `MapEntry` to make the interrupt property optional as needed for.

I agree there's room for debate. We could also use class inheritance to capture the specialization, perhaps. Thanks and looking forward to your next update.

soburi commented:
@mbolivar 

Update done.
Could you take a look, again?

soburi committed:
sys: util: Update document comment

Update document comments for `GET_ARG_N` and `GET_ARGS_LESS_THAN`
to remark the value range.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented:
@mbolivar @rruuaanng @mathieuchopstm 

Can you take a look at it again?

soburi commented on scripts/dts/gen_defines.py:
```diff
@@ -579,6 +580,73 @@ def write_gpio_hogs(node: edtlib.Node) -> None:
+                f'fn(DT_{node.z_path_id}, {prop_id}, {i}, __VA_ARGS__)' for i in range(plen)
+            )
+        )
+
```
If LEN=0, no entries should exist.

mbolivar reviewed (dismissed):
I think this is in good enough shape to merge now and let us iterate on any changes afterwards, in the main branch.

rruuaanng reviewed (dismissed):
Sorry, I'm late.
